cmd/snap-bootstrap/bootstrap/tpm.go

```diff
@@ -0,0 +1,205 @@
+		return err
+	}
+	return nil
+}
+
+// SetShimFiles verifies and sets the list of shim binaries.
+func (t *tpmSupport) SetShimFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.shimFiles = pathList
+	return nil
+}
+
+// SetBootloaderFiles verifies and sets the list of bootloader binaries.
+func (t *tpmSupport) SetBootloaderFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.bootloaderFiles = pathList
+	return nil
+}
+
+// SetKernelFiles verifies and sets the list of kernel binaries.
+func (t *tpmSupport) SetKernelFiles(pathList ...string) error {
+	if err := ensureFilesExist(pathList); err != nil {
+		return err
+	}
+	t.kernelFiles = pathList
+	return nil
+}
+
+func ensureFilesExist(pathList []string) error {
+	for _, p := range pathList {
+		if !osutil.FileExists(p) {
+			return fmt.Errorf("file %s does not exist", p)
+		}
+	}
+	return nil
+}
+
+var (
+	provisionTPM = secboot.ProvisionTPM
+)
+
+// Provision tries to clear and provision the TPM.
+func (t *tpmSupport) Provision() error {
+	if err := provisionTPM(t.tconn, secboot.ProvisionModeFull, t.lockoutAuth); err != nil {
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+
+	return nil
+}
+
+// Close closes the TPM connection.
+func (t *tpmSupport) Close() error {
+	if err := t.tconn.Close(); err != nil {
+		return fmt.Errorf("cannot close TPM connection: %v", err)
+	}
+	return nil
+}
+
+var kernelCmdlines = []string{
+	// run mode
+	"console=ttyS0 console=tty1 panic=-1 systemd.gpt_auto=0 init=/sbin/init snapd_recovery_mode=run",
+	// recover mode
+	"console=ttyS0 console=tty1 panic=-1 systemd.gpt_auto=0 init=/sbin/init snapd_recovery_mode=recover",
+}
+
+var (
+	sealKeyToTPM                  = secboot.SealKeyToTPM
+	addEFISecureBootPolicyProfile = secboot.AddEFISecureBootPolicyProfile
+	addSystemdEFIStubProfile      = secboot.AddSystemdEFIStubProfile
+)
+
+// Seal seals the given key to the TPM and writes the sealed object to a file
+// at the path specified by keyDest. Additional data required for updating the
+// authorization policy is written to a file at the path specified by privDest.
+// This file must live inside an encrypted volume protected by this key.
+func (t *tpmSupport) Seal(key []byte, keyPath, policyUpdatePath string) error {
+	pcrProfile := secboot.NewPCRProtectionProfile()
+
+	// Add EFI secure boot policy profile
+	policyParams := secboot.EFISecureBootPolicyProfileParams{
+		PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+		// TODO:UC20: set SignatureDbUpdateKeystore to support key rotation
+	}
+	for _, shim := range t.shimFiles {
+		s := &secboot.EFIImageLoadEvent{
+			Source: secboot.Firmware,
+			Image:  secboot.FileEFIImage(shim),
+		}
+		for _, bl := range t.bootloaderFiles {
+			g := &secboot.EFIImageLoadEvent{
+				Source: secboot.Shim,
+				Image:  secboot.FileEFIImage(bl),
+			}
+			for _, kernel := range t.kernelFiles {
```

Comment:
I assume this will seal to the signature of the kernel, not the kernel itself? If so, a comment would be nice here.

Reply:
@chrisccoulson what would be the technically accurate description of what we are sealing to, to add as a comment here?

Reply:
We're sealing to the certificate for the key used to verify the kernel.